server/server.go

```diff
@@ -408,6 +408,178 @@ func (s *server) Start(ctx context.Context, req *rpcpb.StartRequest) (*rpcpb.Sta
 	return &rpcpb.StartResponse{ClusterInfo: s.clusterInfo}, nil
 }
 
+func (s *server) DeployBlockchains(ctx context.Context, req *rpcpb.DeployBlockchainsRequest) (*rpcpb.DeployBlockchainsResponse, error) {
+	// if timeout is too small or not set, default to 5-min
+	if deadline, ok := ctx.Deadline(); !ok || time.Until(deadline) < DefaultStartTimeout {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(context.Background(), DefaultStartTimeout)
+		_ = cancel // don't call since "start" is async, "curl" may not specify timeout
+		zap.L().Info("received start request with default timeout", zap.String("timeout", DefaultStartTimeout.String()))
+	} else {
+		zap.L().Info("received start request with existing timeout", zap.String("deadline", deadline.String()))
+	}
+
+	zap.L().Debug("DeployBlockchains")
+	if info := s.getClusterInfo(); info == nil {
+		return nil, ErrNotBootstrapped
+	}
+
+	chainSpecs := []blockchainSpec{}
+	// TODO: fix plugin dir management:
+	// - cannot change plugin dir for nodes that wont be restarted
+	// - can change plugin dir when restating network
+	// - different nodes may have different plugin dirs
+	// - pluging dir shall contain vms for all blockchains
+	if req.GetPluginDir() == "" {
+		if len(req.GetBlockchainSpecs()) > 0 {
+			return nil, ErrPluginDirEmptyButCustomVMsNotEmpty
+		}
+	} else {
+		if len(req.GetBlockchainSpecs()) == 0 {
+			return nil, ErrPluginDirNonEmptyButCustomVMsEmpty
+		}
+		zap.L().Info("non-empty plugin dir", zap.String("plugin-dir", req.GetPluginDir()))
+		for i := range req.GetBlockchainSpecs() {
+			vmName := req.GetBlockchainSpecs()[i].VmName
+			vmGenesisFilePath := req.GetBlockchainSpecs()[i].Genesis
+			zap.L().Info("checking custom VM ID before installation", zap.String("vm-id", vmName))
+			vmID, err := utils.VMID(vmName)
+			if err != nil {
+				zap.L().Warn("failed to convert VM name to VM ID",
+					zap.String("vm-name", vmName),
+					zap.Error(err),
+				)
+				return nil, ErrInvalidVMName
+			}
+			if err := utils.CheckPluginPaths(
+				filepath.Join(req.GetPluginDir(), vmID.String()),
+				vmGenesisFilePath,
+			); err != nil {
+				return nil, err
+			}
+			b, err := os.ReadFile(vmGenesisFilePath)
+			if err != nil {
+				return nil, err
+			}
+			chainSpecs = append(chainSpecs, blockchainSpec{
+				vmName:   vmName,
+				genesis:  b,
+				subnetId: req.GetBlockchainSpecs()[i].SubnetId,
+			})
+		}
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.clusterInfo.CustomVmsHealthy = false
+
+	// start non-blocking to install custom VMs (if applicable)
+	// the user is expected to poll cluster status
+	deployBlockchainsReadyCh := make(chan struct{})
+	go s.network.deployBlockchains(ctx, chainSpecs, deployBlockchainsReadyCh)
+
+	// update cluster info non-blocking
+	// the user is expected to poll this latest information
+	// to decide cluster/subnet readiness
+	go func() {
```

Comment:
Isn't this the exact same 2nd part of start? Maybe we can reuse that so we don't have to replicate code?